### PR TITLE
Add ability to evaluate multiline statements and blocks in one emacs keystroke

### PR DESCRIPTION
**Pull request by obadz**
@7sharp9 some of the changes we had discussed.

### CODE REVIEW

**Comment by 7sharp9** [comment]
@rneatherway Any news on this?  I don't like to see PR's sitting for too long, everyone forgets why things were done the way they were :-)

**Comment by rneatherway** [comment]
I think it's good to merge, but I haven't actually tested it myself yet.

File: emacs/fsharp-mode-indent.el [hunk 1]
```diff
@@ -1476,6 +1477,50 @@ does not include blank lines, comments, or continuation lines."
...
+    (or (fsharp-goto-statement-below) (forward-line 1))
+    (while (and
+            (or
+             (looking-at (concat "[ \t]*\\(" (mapconcat 'identity fsharp-infix-operators-that-begin-lines "\\|") "\\)"))
+             (> (current-indentation) indent))      ; statement indented further
+            (>= (current-indentation) indent)
+            (not (eobp)))
+      (or (fsharp-goto-statement-below) (forward-line 1)))))
+;;    (if (eobp)
+;;        (fsharp-goto-beyond-final-line)))
```
Comment (rneatherway) [inline emacs/fsharp-mode-indent.el]: Remove this commented out code?

File: emacs/fsharp-mode-indent.el [hunk 2]
```diff
@@ -151,7 +151,8 @@ as indentation hints, unless the comment character is in column zero."
 
 
 ;(defconst fsharp-blank-or-comment-re "[ \t]*\\($\\|#\\)"
-(defconst fsharp-blank-or-comment-re "[ \t]*\\(//.*\\)?"
+;(defconst fsharp-blank-or-comment-re "[ \t]*\\(//.*\\)?"
+(defconst fsharp-blank-or-comment-re "[ \t]*\\($\\|//.*\\)"
   "Regular expression matching a blank or comment line.")
```
Comment (rneatherway) [inline emacs/fsharp-mode-indent.el]: Is this used elsewhere?

**Comment by rneatherway** [comment]
@obadz sorry it took me so long to get to this. I think this is a great addition. If you have time could you extend this PR to add this new keybinding to the README and add an integration test like the one at https://github.com/fsharp/fsharpbinding/blob/master/emacs/test/integration-tests.el#L138 ?

**Comment by delexi** [comment]
As this seemed kind of dead, I took the PR for a spin. Is this scenario supposed to work? (with point somewhere on the second or third line)
``` fsharp
let var =
  123
  |> ((-) 456)
```
In my opinion this should work, as a lot of functions are defined this way.

**Comment by obadz** [comment]
@rneatherway sorry I haven't had a chance to revive this, will do shortly.  I have some additional changes that make this a bit better that I will push soon..

@delexi don't remember what the original behavior was intended to do but after I push my new changes the intent is:
* if you hit alt+' on the let var line, it evaluates all 3 lines
* if you hit it on 123, it evaluates 123 |> ((-) 456) (but doesn't assign the var)
* if you hit it on |> ((-) 456) you'll get a parse error

In other words it evaluates the line your on, and all the subsequent line that could not possibly be standalone lines or are indented deeper. Never goes backward.

**Comment by delexi** [comment]
@obadz ah, ok, I see. Looking forward to your contribution :)

**Comment by rneatherway** [comment]
@obadz No problem, will be happy to review when it's ready.

**Comment by obadz** [comment]
Pushed the more recent changes & merged master in.  Let me know what you guys think.

**Comment by rneatherway** [comment]
@obadz looks pretty good. It's a shame that it doesn't really integrate with the other code in `fsharp-mode-indent.el`, but that needs a big overhaul I think. It would be great to have some tests. It doesn't have to be an integration test, we can just test the movement behaviour by stubbing `fsharp-eval-region` (see `fsharp-mode-completion-tests.el`).  I tried on a simple file:

```fsharp
let rec factorial x = 
  if x = 0 then 1 
  else x * (factorial (x - 1))

let f10 = factorial 10

let hidden = 42

let answer = hidden

let laterFunction() = 
  "Not very difficult, is it?"

let sample = 
  laterFunction()
  |> printfn "Got: %s"
```

And I was a bit surprised that on the last block it ends up before the `|>` rather than on the next line.

**Comment by obadz** [comment]
That's because there is no next line to go to.

**Comment by rneatherway** [comment]
It happens even with plenty of empty lines afterwards.

**Comment by obadz** [comment]
Ah I see what you mean.. let me see if I can fix.

**Comment by obadz** [comment]
Moved to https://github.com/fsharp/emacs-fsharp-mode/pull/6
